blocks/datasets/__init__.py

```diff
@@ -0,0 +1,223 @@
+    simultaneously.
+
+    """
+    __metaclass__ = ABCMeta
+    sources = ('features', 'targets')  # Default sources
+
+    def __iter__(self):
+        """Use the default iteration scheme to construct a data stream.
+
+        .. warning::
+
+           A dataset only produces a single data stream using the default
+           iteration scheme, so multiple iterations in parallel are not
+           supported.
+
+        """
+        if not hasattr(self, 'default_scheme'):
+            raise NotImplementedError("Does not provide a default iterator")
+        elif not hasattr(self, 'default_stream'):
+            self.default_stream = DataStream(self, self.default_scheme)
+        return iter(self.default_stream)
+
+    def open(self):
+        """Return the state if the dataset requires one.
+
+        Datasets which e.g. read files from disks require open file
+        handlers, and this sort of stateful information should be handled
+        by the data stream.
+
+        Returns
+        -------
+        state : object
+            An object representing the state of a dataset.
+
+        """
+        pass
+
+    def reset(self, state):
+        """Reset the state in order to start a new epoch.
+
+        This could involve re-opening files, seeking to the beginning, etc.
+
+        Parameters
+        ----------
+        state : object
+            An object representing the state of a dataset
+
+        Returns
+        -------
+        state : object
+            An object representing the state of a dataset.
+
+        Notes
+        -----
+        This is implemented because simply repeatedly calling :meth:`open`
+        could result in file handlers not being closed, and to increase
+        performance by performing a seek-operation instead of re-opening
+        the file.
+
+        """
+        pass
+
+    @abstractmethod
+    def get_data(self, state=None, request=None, sources=None):
+        """Request data from the dataset.
+
+        Parameters
+        ----------
+        state : object, optional
+            The state as returned by the :meth:`open` method. The dataset
+            can use this to e.g. interact with files when needed.
+        request : object, optional
+            If supported, the request for a particular part of the data
+            e.g. the number of examples to return, or the indices of a
+            particular minimbatch of examples.
+        sources : tuple of strings, optional
+            The data sources to return. By default all data sources are
+            returned.
+
+        .. todo::
+
+           A way for the dataset to communicate which kind of requests it
+           accepts, and a way to communicate what kind of request is being
+           sent when supporting multiple.
+
+        Returns
+        -------
+        tuple
+            A tuple of data.
+
+        """
+        raise NotImplementedError
+
+
+class DataStream(object):
```

Comment:
We have a name clash with @rizar #92:
what he calls a data stream is an iterator over iterators,
what you call a data stream is a stateful iterable, each created iterator goes over an epoch.

We should settle on one, I prefer this one - a stream implements the `__iter__` protocol. Especially in extensions that do monitoring it feels weird to code:
```
iter = next(stream)
for batch in iter:
```
I prefer:
```
for batch in stream:
```